sphinx/util/template.py

```diff
@@ -10,16 +10,65 @@
 """
 
 import os
+import sys
+import unicodedata
+from six.moves import reprlib
+
 from jinja2.sandbox import SandboxedEnvironment
 
 from sphinx import package_dir
 from sphinx.jinja2glue import SphinxFileSystemLoader
 
 
+class URepr(reprlib.Repr):
+    """
+    On python 3, repr returns unicode objects, which means that non-ASCII
+    characters are rendered in human readable form.
+
+    This provides a similar facility on python 2.
+
+    Additionally, on python 3, it prefixes unicode repr with a u, such that
+    the returned repr is a valid unicode literal on both python 2 and python 3.
+    """
+
+    # From https://github.com/python/cpython/blob/3.6/Objects/unicodectype.c#L147-L1599
+    nonprintable_categories = ('Cc', 'Cf', 'Cs', 'Co', 'Cn', 'Zl', 'Zp', 'Zs')
+
+    def __init__(self):
+        reprlib.Repr.__init__(self)
+        # disable the reprlib truncation behaviour
+        max_size = 2 ** 128
+        for k in dir(self):
+            if k.startswith('max'):
+                setattr(self, k, max_size)
+
+    if sys.version_info.major >= 3:
+        def repr_str(self, obj, level):
+            return 'u' + repr(obj)
```

Comment:
This is new behaviour, that means that `u`s are inserted into python 3 source code. I'd argue this is useful, as it means a conf file generated in python 3 will work correctly in python 2 (for non-ascii strings).

If you want to keep things as they are, I'm happy to remove this python 3 case